### PR TITLE
Implements `Series.between`

### LINKED ISSUE #1826: Support `Series.between`
Support `Series.between`

### PR DESCRIPTION
**Pull request by gowrijsuria**
## What do these changes do?
This will enable the **between** function for Series.
### Series.between(left, right, inclusive='both')
Returns boolean Series equivalent to left <= series <= right.

This function returns a boolean vector containing True wherever the corresponding Series element is between the boundary values left and right. NA values are treated as False.

Reference in pandas doc: [https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.between.html](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.between.html)

## Related issue number
Fixes #1826

### CODE REVIEW

**Comment by gowrijsuria** [comment]
The test for "test_between_period_values" is failing, for some reason default type "D" for freq is not working. 
TypeError: Cannot interpret 'period[D]' as a data type

File: mars/dataframe/core.py [hunk 1]
```diff
@@ -1268,6 +1269,96 @@ def to_frame(self, name=None):
         name = name or self.name or 0
         return dataframe_from_tensor(self, columns=[name])
 
+    def between(self, left, right, inclusive="both"):
+        """
+        Return boolean Series equivalent to left <= series <= right.
+        This function returns a boolean vector containing `True` wherever the
+        corresponding Series element is between the boundary values `left` and
+        `right`. NA values are treated as `False`.
+        Parameters
```
Comment (qinxuye) [inline mars/dataframe/core.py:1278]: Please pay attention to the blank lines in docs. I will skip to mention the other part in the docs

File: mars/dataframe/core.py [hunk 2]
```diff
@@ -1268,6 +1269,96 @@ def to_frame(self, name=None):
...
+        ----------
+        left : scalar or list-like
+            Left boundary.
+        right : scalar or list-like
+            Right boundary.
+        inclusive : {"both", "neither", "left", "right"}
+            Include boundaries. Whether to set each bound as closed or open.
+            .. versionchanged:: 1.3.0
```
Comment (qinxuye) [inline mars/dataframe/core.py:1286]: Not require versionchanged for Mars.

File: mars/dataframe/core.py [hunk 3]
```diff
@@ -1268,6 +1269,96 @@ def to_frame(self, name=None):
...
+        Series.lt : Less than of series and other.
+        Notes
+        -----
+        This function is equivalent to ``(left <= ser) & (ser <= right)``
+        Examples
+        --------
+        >>> import mars.dataframe as md
+        >>> s = md.Series([2, 0, 4, 8, np.nan])
+        Boundary values are included by default:
+        >>> s.between(1, 4)
```
Comment (qinxuye) [inline mars/dataframe/core.py:1304]: Should be `s.between(1, 4).execute()` instead of `s.execute()`.

File: mars/dataframe/core.py [hunk 4]
```diff
@@ -1268,6 +1269,96 @@ def to_frame(self, name=None):
...
+        >>> s.execute()
+        0     True
+        1    False
+        2     True
+        3    False
+        4    False
+        dtype: bool
+        With `inclusive` set to ``"neither"`` boundary values are excluded:
+        >>> s.between(1, 4, inclusive="neither")
+        >>> s.execute()
```
Comment (qinxuye) [inline mars/dataframe/core.py:1314]: Ditto.

File: mars/dataframe/core.py [hunk 5]
```diff
@@ -1268,6 +1269,96 @@ def to_frame(self, name=None):
...
+        0     True
+        1    False
+        2    False
+        3    False
+        4    False
+        dtype: bool
+        `left` and `right` can be any scalar value:
+        >>> s = md.Series(['Alice', 'Bob', 'Carol', 'Eve'])
+        >>> s.between('Anna', 'Daniel')
+        >>> s.execute()
```
Comment (qinxuye) [inline mars/dataframe/core.py:1324]: Ditto.

File: mars/dataframe/tests/test_core.py [hunk 6]
```diff
@@ -320,3 +320,90 @@ def test_key_value(setup):
...
+
+    result = series.between(left, right).execute().fetch()
+    expected = (series >= left) & (series <= right)
+    expected = expected.execute().fetch()
+    pd.testing.assert_series_equal(result, expected)
+
+
+def test_between_datetime_values(setup):
+    ser = Series(pd.bdate_range("1/1/2000", periods=20).astype(object))
+    ser = ser.execute().fetch()
```
Comment (qinxuye) [inline mars/dataframe/tests/test_core.py:337]: Do not execute here, instead, call `.execute().fetch()` on the `result` variable.

File: mars/dataframe/tests/test_core.py [hunk 7]
```diff
@@ -320,3 +320,90 @@ def test_key_value(setup):
...
+    ser = Series(pd.period_range("2000-01-01", periods=10, freq="D"))
+    left, right = ser[2], ser[7]
+    result = ser.between(left, right).execute().fetch()
+    expected = (ser >= left) & (ser <= right)
+    expected = expected.execute().fetch()
+    pd.testing.assert_series_equal(result, expected)
+
+
+def test_between_inclusive_string(setup):
+    series = Series(pd.date_range("1/1/2000", periods=10))
```
Comment (qinxuye) [inline mars/dataframe/tests/test_core.py:359]: The test can be like.

```
pd_series = pd.xxx
left, right = pd_series[2], pd_series[7]
series = Series(pd_series)
...
result = series.between).to_pandas()
expected = pd_series.between(left, right)
```

The expected part can use pandas directly, the result part in Mars.

**Review by qinxuye — CHANGES_REQUESTED** [review]
I left some comments.

**Comment by qinxuye** [comment]
> The test for "test_between_period_values" is failing, for some reason default type "D" for freq is not working.
> TypeError: Cannot interpret 'period[D]' as a data type

If the test can not be passed, you can uncomment it first, we can solve it later.

**Comment by qinxuye** [comment]
> The test for "test_between_period_values" is failing, for some reason default type "D" for freq is not working.
> TypeError: Cannot interpret 'period[D]' as a data type

Yes, I can reproduce this error described in #2370, and we will address it first, then this PR can merge it to pass the ut.

**Comment by wjsi** [comment]
Two suggestions for tests.

1. Not all pandas tests are needed for Mars as Mars rely on pandas for computation. You may only pick cases covering all your branches.
2. Try creating source data with multiple chunks, for instance, `df = DataFrame(raw, chunk_size=5)`, to make cases close to real scenarios.

**Comment by qinxuye** [comment]
> Two suggestions for tests.
> 
> 1. Not all pandas tests are needed for Mars as Mars rely on pandas for computation, and it costs too much time running all cases. You may only pick cases covering all your branches.
> 2. Try creating source data with multiple chunks, for instance, `df = DataFrame(raw, chunk_size=5)`, to make cases close to real scenarios.

About this, @gowrijsuria you can refer to the changes in #2367 .

File: mars/dataframe/tests/test_core.py (hunk 6 above)
Comment (gowrijsuria) [reply mars/dataframe/tests/test_core.py:337]: If I remove line 337, this error appears.
>       ser[::2] = np.nan
E       TypeError: 'Series' object does not support item assignment
Therefore, I have kept line 337.

Comment (qinxuye) [reply mars/dataframe/tests/test_core.py:337]: You can create a pandas array, after item assignment, create Mars series.

```
raw = pd.bdate_range(xxx)
raw[::2] = np.nan
ser = Series(raw)
...
```

File: mars/dataframe/tests/test_core.py [hunk 8]
```diff
@@ -320,3 +320,92 @@ def test_key_value(setup):
...
+
+
+def test_between_inclusive_warning(setup):
+    pd_series = pd.Series(pd.date_range("1/1/2000", periods=10))
+    pd_left, pd_right = pd_series[2], pd_series[7]
+    series = Series(pd_series, chunk_size=5)
+    left, right = series[2], series[7]
+    with pd.testing.assert_produces_warning(FutureWarning):
+        result = series.between(left, right, inclusive=False).execute().fetch()
+        expected = (pd_series > pd_left) & (pd_series < pd_right)
```
Comment (wjsi) [inline mars/dataframe/tests/test_core.py:406]: Does not need to execute. Just test if a warning is produced.

File: mars/dataframe/tests/test_core.py [hunk 9]
```diff
@@ -320,3 +320,92 @@ def test_key_value(setup):
...
+        pd.testing.assert_series_equal(result, expected)
+    with pd.testing.assert_produces_warning(FutureWarning):
+        result = series.between(left, right, inclusive=True).execute().fetch()
+        expected = (pd_series >= pd_left) & (pd_series <= pd_right)
```
Comment (wjsi) [inline mars/dataframe/tests/test_core.py:410]: Ditto.

File: mars/dataframe/tests/test_core.py [hunk 10]
```diff
@@ -320,3 +320,92 @@ def test_key_value(setup):
 
     result = idx.values.execute().fetch()
     np.testing.assert_array_equal(result, raw.values)
+
+
+def test_between(setup):
+    pd_series = pd.Series(pd.date_range("1/1/2000", periods=10))
+    pd_left, pd_right = pd_series[2], pd_series[7]
+    expected = (pd_series >= pd_left) & (pd_series <= pd_right)
```
Comment (wjsi) [inline mars/dataframe/tests/test_core.py:328]: Also use `between` in pandas code to make sure behaviors between pandas and Mars are consistent.

**Review by wjsi — CHANGES_REQUESTED** [review]

File: mars/dataframe/tests/test_core.py (hunk 8 above)
Comment (gowrijsuria) [reply mars/dataframe/tests/test_core.py:406]: So I need to just remove lines 406 and 407?

Comment (qinxuye) [reply mars/dataframe/tests/test_core.py:406]: Guess so, the warning is gonna produced even without `execute`.

Comment (gowrijsuria) [reply mars/dataframe/tests/test_core.py]: done

File: mars/dataframe/tests/test_core.py [hunk 11]
```diff
@@ -320,3 +320,88 @@ def test_key_value(setup):
...
+    result = result.execute().fetch()
+    pd.testing.assert_series_equal(result, expected)
+
+    result = series[series.between(series[3], series[17], inclusive="neither")]
+    result = result.execute().fetch()
+    expected = pd_series[5:16].dropna()
+    pd.testing.assert_series_equal(result, expected)
+
+
+# def test_between_period_values(setup):
```
Comment (qinxuye) [inline mars/dataframe/tests/test_core.py:352]: This test should be able to run after merging the master branch.

Comment (gowrijsuria) [reply mars/dataframe/tests/test_core.py:352]: ok , I will uncomment it.

File: mars/dataframe/tests/test_core.py [hunk 12]
```diff
@@ -254,3 +254,88 @@ def test_key_value(setup):
...
+    result = series[series.between(series[3], series[17], inclusive="neither")]
+    result = result.execute().fetch()
+    expected = pd_series[5:16].dropna()
+    pd.testing.assert_series_equal(result, expected)
+
+
+def test_between_period_values(setup):
+    pd_series = pd.Series(pd.period_range("2000-01-01", periods=10, freq="D"))
+    pd_left, pd_right = pd_series[2], pd_series[7]
+    expected = (pd_series >= pd_left) & (pd_series <= pd_right)
```
Comment (wjsi) [inline mars/dataframe/tests/test_core.py:289]: Use `between` in pandas to calculate expected result.

Comment (gowrijsuria) [reply mars/dataframe/tests/test_core.py]: Done

**Comment by gowrijsuria** [comment]
Some checks are failing. How to fix them?

**Comment by qinxuye** [comment]
> Some checks are failing. How to fix them?

Only this is relevant to this PR: https://dev.azure.com/mars-project/mars/_build/results?buildId=1122&view=logs&jobId=c2eb7656-25f5-53fe-1976-0814402b63cf&j=c2eb7656-25f5-53fe-1976-0814402b63cf&t=b59d65c4-93af-5978-6402-a4f708982468

Maybe try `pd._testing.assert_produces_warning` instead.

**Comment by gowrijsuria** [comment]
Replacing with " **pd._testing.assert_produces_warning(FutureWarning ,"Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.")** ", gives this error:
AssertionError: Did not see expected warning of class 'FutureWarning'

**Comment by wjsi** [comment]
Try using https://docs.python.org/3/library/warnings.html#testing-warnings instead. `_testing` is not a public module of pandas.

**Comment by qinxuye** [comment]
> Replacing with " **pd._testing.assert_produces_warning(FutureWarning ,"Boolean inputs to the `inclusive` argument are deprecated in favour of `both` or `neither`.")** ", gives this error:
> AssertionError: Did not see expected warning of class 'FutureWarning'

I tried to pull the code, and I found `pd._testing.assert_produces_warning` can work, but like @wjsi said, we should not rely on internal apis of pandas, thus I suggest we can use pytest as well.

Below code should be workable.

```
with pytest.warns(FutureWarning):
        series.between(left, right, inclusive=False)
    with pytest.warns(FutureWarning):
        series.between(left, right, inclusive=True)
```

**Review by qinxuye — APPROVED** [review]
LGTM

**Review by wjsi — APPROVED** [review]
LGTM

**Comment by qinxuye** [comment]
Congrats, look forward to seeing more contributions from you.

**Comment by gowrijsuria** [comment]
Thank you @qinxuye @wjsi !

**Pull request merged**